service/matching/version_graph.go

```diff
@@ -0,0 +1,256 @@
+					PreviousCompatible:   curDefault,
+					PreviousIncompatible: curDefault.PreviousIncompatible,
+				}
+			} else {
+				// Otherwise, set the previous incompatible version to the current default.
+				existingData.CurrentDefault = &taskqueuepb.VersionIdNode{
+					Version:              req.VersionId,
+					PreviousCompatible:   nil,
+					PreviousIncompatible: curDefault,
+				}
+			}
+		} else {
+			existingData.CurrentDefault = &taskqueuepb.VersionIdNode{
+				Version:              req.VersionId,
+				PreviousCompatible:   nil,
+				PreviousIncompatible: nil,
+			}
+		}
+	} else {
+		if req.GetPreviousCompatible() != nil {
+			prevCompat, indexInCompatLeaves := findCompatibleNode(existingData, req.GetPreviousCompatible())
+			if prevCompat != nil {
+				newNode := &taskqueuepb.VersionIdNode{
+					Version:              req.VersionId,
+					PreviousCompatible:   prevCompat,
+					PreviousIncompatible: nil,
+				}
+				if indexInCompatLeaves >= 0 {
+					existingData.CompatibleLeaves[indexInCompatLeaves] = newNode
+				} else {
+					existingData.CompatibleLeaves = append(existingData.CompatibleLeaves, newNode)
+				}
+			} else {
+				return serviceerror.NewNotFound(
+					fmt.Sprintf("previous compatible version %v not found", req.GetPreviousCompatible()))
+			}
+		} else {
+			// Check if the version is already a default, and remove it from being one if it is.
+			curDefault := existingData.GetCurrentDefault()
+			if curDefault.GetVersion().Equal(req.GetVersionId()) {
+				existingData.CurrentDefault = nil
+				if curDefault.GetPreviousCompatible() != nil {
+					existingData.CurrentDefault = curDefault.GetPreviousCompatible()
+				} else if curDefault.GetPreviousIncompatible() != nil {
+					existingData.CurrentDefault = curDefault.GetPreviousIncompatible()
+				}
+				return nil
+			}
+			// Check if it's a compatible leaf, and remove it from being one if it is.
+			for i, def := range existingData.GetCompatibleLeaves() {
+				if def.GetVersion().Equal(req.GetVersionId()) {
+					existingData.CompatibleLeaves =
+						append(existingData.CompatibleLeaves[:i], existingData.CompatibleLeaves[i+1:]...)
+					if def.GetPreviousCompatible() != nil {
+						existingData.CompatibleLeaves =
+							append(existingData.CompatibleLeaves, def.GetPreviousCompatible())
+					}
+					return nil
+				}
+			}
+			return serviceerror.NewInvalidArgument(
+				"requests to update build id ordering cannot create a new non-default version with no links")
+		}
+	}
+	return nil
+}
+
+// Finds the node that the provided version should point at, given that it says it's compatible with the provided
+// version. Note that this does not necessary mean *that* node. If the version being targeted as compatible has nodes
+// which already point at it as their previous compatible version, that chain will be followed out to the leaf, which
+// will be returned.
+func findCompatibleNode(
+	existingData *persistence.VersioningData,
+	versionId *taskqueuepb.VersionId,
```

Comment:
Can you confirm you never expect the caller to pass in nil for versionID here? No need to add code asserting it, just want to confirm because it can change up the `Equal` logic downstream.

Reply:
Yup, will never be nil.